### PR TITLE
chore: add audit to check if an EOL is known

### PR DESCRIPTION
**Pull request by SMillerDev**
- [ ] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----
This should give us some warning if software has a known EOL date

### CODE REVIEW

**Comment by SMillerDev** [comment]
```console
cake
  * Product is EOL
isc-dhcp
  * Product is EOL since 2022-10-05
prometheus
  * Product is EOL since 2023-12-28
qt@5
  * Product is EOL since 2023-05-26
spark
  * Product is EOL since 2014-09-03
```
Cake probably redirects to CakePHP.
Prometheus shot themselves in the foot by not releasing a new version before letting the old one expire.
Not sure what the state of QT is.
Spark thinks it's apache-spark, so that's not great.

**Comment by Bo98** [comment]
> Not sure what the state of QT is.

Qt5 isn't fully accurate. There's releases planned until 2025: https://wiki.qt.io/Qt_5.15_Release. And even then there'll be a further year due to the open source release delay that may not be reflected in the EOL date.

[EOL page](https://endoflife.date/qt) calls it commercial support, which is correct but that commercial support started in 2021 rather than 2023 and the agreement made with KDE was that all commercial updates become free after one year. KDE also host backports should anything urgent surface in that delay.

**Comment by Bo98** [comment]
> Prometheus shot themselves in the foot by not releasing a new version before letting the old one expire.

Looking at the history of EOL dates, this seems to happen very regularly so the dates probably can't be trusted for this one.

The wording is: 

> Every 6 weeks, a new Prometheus minor release cycle begins. After those 6 weeks, minor releases generally no longer receive bugfixes.

I think what they really mean is "when the next minor release happens the previous won't receive bug fixes", which is fairly normal.

File: Library/Homebrew/formula_auditor.rb
Comment (Bo98) [inline Library/Homebrew/formula_auditor.rb:594]:
```suggestion
      return if formula.deprecated? || formula.disabled?
      return if @core_tap
```

given different taps may have different policies (e.g. taps that do `brew extract`).

File: Library/Homebrew/formula_auditor.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_auditor.rb:609]: Is there a URL we can supply for this to the endoflife.date page?

File: Library/Homebrew/utils/shared_audits.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/shared_audits.rb:20]: Wondering if we want some `opoo` or `odebug` here so e.g. a change to the API doesn't result in us silently not noticing this doesn't work any more?

**Review by MikeMcQuaid — COMMENTED** [review]
Great work so far @SMillerDev, really good idea.

File: Library/Homebrew/formula_auditor.rb
Comment (SMillerDev) [reply Library/Homebrew/formula_auditor.rb:609]: Something like this?
```suggestion
        problem "Product is EOL, see https://endoflife.date/#{name}"
```

File: Library/Homebrew/utils/shared_audits.rb
Comment (SMillerDev) [reply Library/Homebrew/utils/shared_audits.rb:20]: I can do that, what do you want to error on?

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/shared_audits.rb:20]: @SMillerDev Feels like should definitely error on JSON failing to parse (which I think it maybe already does?). Not sure if the expected failure code is an unsuccessful status code or the product not found. Would be good to fail on whatever is not expected.

Comment (SMillerDev) [reply Library/Homebrew/utils/shared_audits.rb:20]: Yeah, JSON parsing will fail. I believe the "Not Found" failure is both a 404 and a message actually.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/shared_audits.rb:20]: Yeh, can probably skip the message check then I think and fail on anything except a `404` status code?

**Comment by cho-m** [comment]
> Spark thinks it's apache-spark, so that's not great.

Probably going to need an audit exception to handle this scenario. May also want to handle reverse of mapping Homebrew's name to a name that endoflife.date supports.

File: Library/Homebrew/utils/shared_audits.rb
Comment (SMillerDev) [reply Library/Homebrew/utils/shared_audits.rb:20]: Do we have any reference for checking status codes? I can't find any.

**Comment by SMillerDev** [comment]
Added the exception, I punted the reverse mapping since it might also be good to just add an alias on the endoflife.date side if we want that.

**Comment by cho-m** [comment]
> I punted the reverse mapping since it might also be good to just add an alias on the endoflife.date side if we want that.

That's fine with me. An odd formula would be `openjdk` as I don't know if we want to point that to `java` which is alias to `oracle-jdk` (commercial).

File: Library/Homebrew/utils/shared_audits.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/utils/shared_audits.rb:20]: `curl_check_http_content` and `curl_http_content_headers_and_checksum` and `parse_curl_output` in `utils/curl` should provide some examples here.

Feel free to say "this is too involved, I can't be bothered" and that's fine 😁

File: Library/Homebrew/formula_auditor.rb
Comment (cho-m) [reply Library/Homebrew/formula_auditor.rb:594]: @SMillerDev I think you lost this change on last push

File: Library/Homebrew/utils/shared_audits.rb
Comment (SMillerDev) [reply Library/Homebrew/utils/shared_audits.rb:20]: Yeah, that looks a lot more complicated than parsing JSON. Gonna keep it as-is

**Comment by Bo98** [comment]
Before shipping this also needs a `returns unless @online`. It's ok for now for testing but it can't ship as a tap-wide check since it'll make it take 40 minutes.

**Comment by SMillerDev** [comment]
Their server is GitHub page so I'm sure it'll be fine. But I'll as the flag.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good to me to merge, thanks @SMillerDev! @SMillerDev can you self-merge and be around to revert/fix forward if needed in case this goes 💥 on homebrew-core.

**Comment by SMillerDev** [comment]
Sure, don't have any approvals on the required core change though.

**Comment by MikeMcQuaid** [comment]
> Sure, don't have any approvals on the required core change though.

@SMillerDev whoops: you do now ✅ 😅

**Comment by Bo98** [comment]
Looks like the `return unless @core_tap` got dropped. Was that intentional?

**Comment by SMillerDev** [comment]
Not sure what happened there. I'll add it in the other PR I have open

**Pull request merged**
